core/src/main/kotlin/org/kkafka/producer/send.kt

```diff
@@ -0,0 +1,16 @@
+package org.kkafka.producer
+
+import org.apache.kafka.clients.producer.KafkaProducer
+import org.apache.kafka.clients.producer.ProducerRecord
+import org.apache.kafka.clients.producer.RecordMetadata
+import kotlin.Result.Companion.failure
+import kotlin.Result.Companion.success
+import kotlin.coroutines.suspendCoroutine
+
+public suspend fun <K, V> KafkaProducer<K, V>.sendSuspending(record: ProducerRecord<K, V>): RecordMetadata =
+    suspendCoroutine { continuation ->
+        send(record) { metadata: RecordMetadata?, exception: Exception? ->
+            val res = if (exception != null) failure(exception) else success(metadata!!)
```

Comment:
we really really cannot throw in here, so maybe it would be safer you return an empty `RecordMetadata` if the one give to us is null?
Otherwise we can investigate (in kafka docs) whether this can ever be null if exception is null..

Reply:
Technically there should only be two options
- exception == null && metadata != null
- exception != null && metadata == null
But the doc is not 100% explicit on that so I'll take that into account

Reply:
Actually I seem to remember that RecordMetadata can't be null and if there is an exception it will be initialised with all -1s

Reply:
https://kafka.apache.org/10/javadoc/org/apache/kafka/clients/producer/Callback.html

Should be good like as it is, what do you think?

Reply:
yeah looks good you're right

it's just that if these throw (rather than resumeWith success or failure) then the dispatcher will just get stuck rather than crash the program which is not good :c at all

Reply:
and that worries me the bugs would be horrible to find